server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java

```diff
@@ -88,4 +141,355 @@ public void testRecoverStateUpdateTask() throws Exception {
+        assertClusterStateBlocks(true);
+        assertThat(gatewayService.currentPendingStateRecovery, sameInstance(pendingStateRecoveryOfTerm1));
+
+        // Will run recover again for a newer term
+        final ClusterState clusterStateOfTerm2 = ClusterState.builder(initialState)
+            .metadata(Metadata.builder().coordinationMetadata(CoordinationMetadata.builder().term(2).build()).build())
+            .build();
+        setClusterStateTaskQueue.submitTask(randomAlphaOfLength(5), new SetClusterStateTask(clusterStateOfTerm2), null);
+        assertThat(deterministicTaskQueue.hasRunnableTasks(), is(true));
+        deterministicTaskQueue.runAllRunnableTasks();
+        assertClusterStateBlocks(false);
+        assertThat(rerouteCount.get(), equalTo(2));
+        assertThat(gatewayService.currentPendingStateRecovery, not(sameInstance(pendingStateRecoveryOfTerm1)));
+
+        // Never ran any scheduled task since recovery is immediate
+        assertThat(deterministicTaskQueue.hasDeferredTasks(), is(false));
+        assertThat(deterministicTaskQueue.getCurrentTimeMillis(), equalTo(initialTimeInMillis));
+    }
+
+    public void testScheduledRecovery() {
+        final Settings.Builder settingsBuilder = Settings.builder();
+        final int expectedNumberOfDataNodes = randomIntBetween(2, 5);
+        settingsBuilder.put(EXPECTED_DATA_NODES_SETTING.getKey(), expectedNumberOfDataNodes);
+        final boolean hasRecoverAfterTime = randomBoolean();
+        if (hasRecoverAfterTime) {
+            settingsBuilder.put(RECOVER_AFTER_TIME_SETTING.getKey(), TimeValue.timeValueMinutes(10));
+        }
+
+        final ClusterState initialState = buildClusterState(1, 0);
+        final GatewayService gatewayService = createService(settingsBuilder, initialState);
+        assertClusterStateBlocks(true);
+
+        // The 1st scheduled recovery for the initial term
+        final long initialTimeInMillis = deterministicTaskQueue.getCurrentTimeMillis();
+        final var setClusterStateTaskQueue = createSetClusterStateTaskQueue(clusterService);
+        final ClusterState clusterStateOfTerm1 = incrementTerm(initialState);
+        setClusterStateTaskQueue.submitTask(randomAlphaOfLength(5), new SetClusterStateTask(clusterStateOfTerm1), null);
+        deterministicTaskQueue.runAllRunnableTasks(); // publish cluster state term change
+        // recovery is scheduled but has not run yet
+        assertThat(deterministicTaskQueue.hasDeferredTasks(), is(true));
+        assertClusterStateBlocks(true);
+        assertThat(rerouteCount.get(), equalTo(0));
+        final GatewayService.PendingStateRecovery pendingStateRecoveryOfInitialTerm = gatewayService.currentPendingStateRecovery;
+        assertThat(pendingStateRecoveryOfInitialTerm, notNullValue());
+
+        // The 1st scheduled recovery may or may not run depend on what happens to the cluster next
+        final int caseNo = randomIntBetween(0, 2);
+        switch (caseNo) {
```

Comment:
Could we test these cases as separate tests (so they all run every time)?

Reply:
Yeah that's fair. I pushed 19f89bac90000265796a6b551e56049cb5cea881